compiler/generator_php.c

```diff
@@ -0,0 +1,1406 @@
+                if (j < 0 || j > (int)(sizeof(g->B) / sizeof(g->B[0])))
+                    goto invalid_escape2;
+                write_s(g, g->B[j]);
+                continue;
+            }
+            case 'I': {
+                int j = input[i++] - '0';
+                if (j < 0 || j > (int)(sizeof(g->I) / sizeof(g->I[0])))
+                    goto invalid_escape2;
+                write_int(g, g->I[j]);
+                continue;
+            }
+            case 'V':
+                write_varref(g, p->name);
+                continue;
+            case 'W':
+                write_varname(g, p->name);
+                continue;
+            case 'L':
+                write_literal_string(g, p->literalstring);
+                continue;
+            case '+': g->margin++; continue;
+            case '-': g->margin--; continue;
+            case 'n': write_string(g, g->options->name); continue;
+            case 'P': write_string(g, g->options->parent_class_name); continue;
+            case 'C': { // Constant.
+                w(g, "/* TODO */ const"); // ??
+                continue;
+            }
+            case 'D': { // Declare variable.
+                w(g, "/* TODO ?*/" ); // ??
+                continue;
+            }
+            default:
+                printf("Invalid escape sequence ~%c in writef(g, \"%s\", p)\n",
+                       ch, input);
+                exit(1);
+            invalid_escape2:
+                printf("Invalid escape sequence ~%c%c in writef(g, \"%s\", p)\n",
+                       ch, input[i - 1], input);
+                exit(1);
+        }
+    }
+}
+
+static void w(struct generator * g, const char * s) {
+    writef(g, s, NULL);
+}
+
+static void generate_AE(struct generator * g, struct node * p) {
+    const char * s;
+    switch (p->type) {
+        case c_name:
+            write_string(g,"$this->");
+            write_varref(g, p->name); break;
+        case c_number:
+            write_int(g, p->number); break;
+        case c_maxint:
+            write_string(g, "(-1>>>1)/*FIXME*/"); break;
+        case c_minint:
+            write_string(g, "(~(-1>>>1))/*FIXME*/"); break;
+        case c_neg:
+            write_char(g, '-'); generate_AE(g, p->right); break;
+        case c_multiply:
+            s = " * "; goto label0;
+        case c_plus:
+            s = " + "; goto label0;
+        case c_minus:
+            s = " - ";
+        label0:
+            write_char(g, '('); generate_AE(g, p->left);
+            write_string(g, s); generate_AE(g, p->right); write_char(g, ')'); break;
+        case c_divide:
+            /* Snowball specifies integer division with semantics matching C,
+             * so we need to use `Math.trunc(x/y)` here.
+             */
+            write_string(g, "$this->trunc_FIXME(");
+            generate_AE(g, p->left);
+            write_string(g, " / ");
+            generate_AE(g, p->right);
+            write_char(g, ')');
```

Comment:
Looks like PHP's `intdiv()` has the correct semantics: https://www.php.net/manual/en/function.intdiv.php

Reply:
Implemented in 4e78723